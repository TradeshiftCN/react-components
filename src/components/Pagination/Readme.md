### Basic Usage

```jsx
<Pagination
	pages={10}
	defaultPage={2}
	onChange={page => {
		console.log(page);
	}}
/>
```

### Two-way binding

#### Fixed
```jsx
<Pagination page={4} pages={13} />
```
#### Update with callback
```jsx
initialState = { page: 3 };
<Pagination
	page={state.page}
	pages={13}
	onChange={page => {
		setState({ page: page });
	}}
/>;
```
