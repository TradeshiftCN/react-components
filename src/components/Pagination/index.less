@import '../../style/variables/index';

.ts-rc-pagination {
	.button-color(@color) {
		color: @color;
		> svg {
			fill: @color;
		}
	}
	font-size: @ts-fontsize;
	button {
		width: @ts-unit-double;
		height: @ts-unit-double;

		> * {
			vertical-align: middle;
		}

		&:disabled {
			.button-color(@ts-color-gray);
		}

		&:not(:disabled) {
			.button-color(@ts-color-blue);
			&:hover {
				.button-color(@ts-color-blue-dark);
				text-decoration: none;
			}
			&.ts-rc-pagination-btn--active {
				color: @ts-color-black;
			}
		}
	}

	&-pages {
		display: inline-block;
		font-weight: @ts-fontweight-semibold;
		&--less button:first-child span {
			direction: rtl;
		}
		&--less button:first-child span::before,
		&--more button:last-child span::after {
			content: '\2026';
			// taking '...' out of flow without setting left/top
			// it'll follow previous element but text-align: center won't calc it's width
			position: absolute;
			margin-top: 2px;
			font-size: @ts-fontsize-mini;
			font-weight: @ts-fontweight;
		}
	}
}
