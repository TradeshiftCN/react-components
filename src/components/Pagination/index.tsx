import React, { Component } from 'react';
import cx from 'classnames';
import _ from 'lodash';

type PaginationProps = {
	prefixCls?: string;
	defaultPage?: number;
	page?: number;
	pages?: number;
	className?: string;
	onChange?(page: number): void;
};

type PaginationState = {
	internalPage?: number;
	prevPropsPage?: number;
	numbers: number[];
};

class Pagination extends Component<PaginationProps, PaginationState> {
	static defaultProps = {
		prefixCls: 'ts-rc-pagination'
	};

	state: PaginationState = {
		internalPage: this.props.defaultPage,
		prevPropsPage: this.props.page,
		numbers: []
	};

	render() {
		const { prefixCls, pages, className } = this.props;
		const { numbers } = this.state;
		const current = this.page;

		if (!numbers.length) {
			return null;
		}

		const hasPrev = current !== 1;
		const hasNext = current !== pages;
		const hasHiddenPrev = numbers[0] !== 1;
		const hasHiddenNext = numbers[numbers.length - 1] !== pages;

		return (
			<div className={cx(prefixCls, className)}>
				<button type="button" disabled={!hasPrev} onClick={this.goFirst}>
					<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24">
						<path d="M6 6h2v12H6zm3.5 6l8.5 6V6z"></path>
					</svg>
				</button>
				<button type="button" disabled={!hasPrev} onClick={this.goPrev}>
					<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24">
						<path d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"></path>
					</svg>
				</button>
				<span
					className={cx(`${prefixCls}-pages`, {
						[`${prefixCls}-pages--more`]: hasHiddenNext,
						[`${prefixCls}-pages--less`]: hasHiddenPrev
					})}
				>
					{numbers.map(page => (
						<button
							key={page}
							type="button"
							onClick={() => this.setCurrent(page)}
							className={cx({
								[`${prefixCls}-btn--active`]: current === page
							})}
						>
							<span>{page}</span>
						</button>
					))}
				</span>
				<button type="button" disabled={!hasNext} onClick={this.goNext}>
					<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24">
						<path d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"></path>
					</svg>
				</button>
				<button type="button" disabled={!hasNext} onClick={this.goLast}>
					<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24">
						<path d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"></path>
					</svg>
				</button>
			</div>
		);
	}

	componentDidMount() {
		this.updateNumbers();
	}

	componentDidUpdate(prevProps: PaginationProps) {
		if (prevProps.pages !== this.props.pages) {
			this.updateNumbers();
		}
	}

	private get page(): number {
		const { page } = this.props;
		return _.isNumber(page) ? page : this.state.internalPage || 0;
	}

	private updateNumbers() {
		const { pages } = this.props;
		if (!pages || pages < 1) {
			this.setState({ numbers: [] });
			return;
		}

		// init numbers
		const numbers = this.generateNumbers({ start: 1 });
		if (numbers.indexOf(this.page) === -1) {
			this.setCurrent(numbers[0]);
		}
	}

	private generateNumbers({ start, end }: { start?: number; end?: number }) {
		const { pages } = this.props;
		const numbers: number[] = [];

		if (!start && !end) {
			start = 1;
		}

		if (start) {
			for (let i = start; i < start + 5 && i <= pages!; i++) {
				numbers.push(i);
			}
		}

		if (end) {
			for (let i = end; i > end - 5 && i >= 1; i--) {
				numbers.push(i);
			}
			numbers.reverse();
		}
		this.setState({ numbers });

		return numbers;
	}

	private setCurrent(current: number): Promise<boolean> {
		const page = this.page;
		return new Promise(resolve => {
			if (this.props.onChange) {
				this.props.onChange(current);
			}
			this.setState(
				{
					internalPage: current
				},
				() => {
					resolve(page !== this.page);
				}
			);
		});
	}

	private goFirst = async () => {
		const changed = await this.setCurrent(1);

		if (changed) {
			this.generateNumbers({ start: 1 });
		}
	};

	private goPrev = async () => {
		const { numbers } = this.state;
		const current = this.page;
		const changed = await this.setCurrent(this.page - 1);
		if (changed && numbers.indexOf(current - 1) === -1) {
			this.generateNumbers({ start: numbers[0] - 1 });
		}
	};

	private goNext = async () => {
		const { numbers } = this.state;
		const current = this.page;
		const changed = await this.setCurrent(current + 1);
		if (changed && numbers.indexOf(current + 1) === -1) {
			this.generateNumbers({ start: numbers[0] + 1 });
		}
	};

	private goLast = async () => {
		const changed = await this.setCurrent(this.props.pages!);
		if (changed) {
			this.generateNumbers({ end: this.props.pages });
		}
	};
}

export default Pagination;
