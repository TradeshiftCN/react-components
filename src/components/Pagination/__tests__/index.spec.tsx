import React from 'react';
import { shallow } from 'enzyme';
import Pagination from '../';

describe('Pagination', () => {
	it('basic', () => {
		const wrapper = shallow(<Pagination></Pagination>);
		// no pages
		expect(wrapper.isEmptyRender()).toBe(true);

		wrapper.setProps({
			pages: 3
		});
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({
			pages: 4,
			page: 2
		});
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({
			pages: 9,
			page: 3
		});
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({
			pages: 9,
			page: 7
		});
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({
			pages: 9,
			page: 9
		});
		expect(wrapper).toMatchSnapshot();

		wrapper.setProps({ pages: undefined });
		expect(wrapper.isEmptyRender()).toBe(true);
	});

	it('event', () => {
		const onChange = jest.fn();
		const wrapper = shallow(<Pagination pages={8} onChange={onChange}></Pagination>);

		// if missing page, Pagination will set default value
		expect(onChange.mock.calls[0][0]).toBe(1);

		onChange.mockClear();
		wrapper
			.find('button')
			.at(7)
			.simulate('click');
		expect(onChange.mock.calls[0][0]).toBe(2);

		onChange.mockClear();
		wrapper
			.find('button')
			.last()
			.simulate('click');

		expect(onChange.mock.calls[0][0]).toBe(8);

		onChange.mockClear();
		wrapper
			.find('button')
			.at(1)
			.simulate('click');
		expect(onChange.mock.calls[0][0]).toBe(7);

		onChange.mockClear();
		wrapper
			.find('button')
			.first()
			.simulate('click');
		expect(onChange.mock.calls[0][0]).toBe(1);

		onChange.mockClear();
		wrapper
			.find('button')
			.at(4)
			.simulate('click');
		expect(onChange.mock.calls[0][0]).toBe(3);
	});

	it('pass props', () => {
		const onChange = jest.fn((page: number) => {
			wrapper.setProps({
				page
			});
		});
		const wrapper = shallow(<Pagination pages={12} onChange={onChange} page={5}></Pagination>);

		expect(wrapper.find('.ts-rc-pagination-btn--active').text()).toBe('5');
		wrapper
			.find('button')
			.at(7)
			.simulate('click');
		expect(onChange.mock.calls[0][0]).toBe(6);
	});
});
