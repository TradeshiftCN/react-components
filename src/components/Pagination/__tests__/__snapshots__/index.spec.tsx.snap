// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination basic 1`] = `
<div
  className="ts-rc-pagination"
>
  <button
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
      />
    </svg>
  </button>
  <button
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
      />
    </svg>
  </button>
  <span
    className="ts-rc-pagination-pages"
  >
    <button
      className="ts-rc-pagination-btn--active"
      key="1"
      onClick={[Function]}
      type="button"
    >
      <span>
        1
      </span>
    </button>
    <button
      className=""
      key="2"
      onClick={[Function]}
      type="button"
    >
      <span>
        2
      </span>
    </button>
    <button
      className=""
      key="3"
      onClick={[Function]}
      type="button"
    >
      <span>
        3
      </span>
    </button>
  </span>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination basic 2`] = `
<div
  className="ts-rc-pagination"
>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
      />
    </svg>
  </button>
  <span
    className="ts-rc-pagination-pages"
  >
    <button
      className=""
      key="1"
      onClick={[Function]}
      type="button"
    >
      <span>
        1
      </span>
    </button>
    <button
      className=""
      key="2"
      onClick={[Function]}
      type="button"
    >
      <span>
        2
      </span>
    </button>
    <button
      className="ts-rc-pagination-btn--active"
      key="3"
      onClick={[Function]}
      type="button"
    >
      <span>
        3
      </span>
    </button>
    <button
      className=""
      key="4"
      onClick={[Function]}
      type="button"
    >
      <span>
        4
      </span>
    </button>
  </span>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination basic 3`] = `
<div
  className="ts-rc-pagination"
>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
      />
    </svg>
  </button>
  <span
    className="ts-rc-pagination-pages ts-rc-pagination-pages--more"
  >
    <button
      className=""
      key="1"
      onClick={[Function]}
      type="button"
    >
      <span>
        1
      </span>
    </button>
    <button
      className=""
      key="2"
      onClick={[Function]}
      type="button"
    >
      <span>
        2
      </span>
    </button>
    <button
      className="ts-rc-pagination-btn--active"
      key="3"
      onClick={[Function]}
      type="button"
    >
      <span>
        3
      </span>
    </button>
    <button
      className=""
      key="4"
      onClick={[Function]}
      type="button"
    >
      <span>
        4
      </span>
    </button>
    <button
      className=""
      key="5"
      onClick={[Function]}
      type="button"
    >
      <span>
        5
      </span>
    </button>
  </span>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination basic 4`] = `
<div
  className="ts-rc-pagination"
>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
      />
    </svg>
  </button>
  <span
    className="ts-rc-pagination-pages ts-rc-pagination-pages--more"
  >
    <button
      className=""
      key="1"
      onClick={[Function]}
      type="button"
    >
      <span>
        1
      </span>
    </button>
    <button
      className=""
      key="2"
      onClick={[Function]}
      type="button"
    >
      <span>
        2
      </span>
    </button>
    <button
      className=""
      key="3"
      onClick={[Function]}
      type="button"
    >
      <span>
        3
      </span>
    </button>
    <button
      className=""
      key="4"
      onClick={[Function]}
      type="button"
    >
      <span>
        4
      </span>
    </button>
    <button
      className=""
      key="5"
      onClick={[Function]}
      type="button"
    >
      <span>
        5
      </span>
    </button>
  </span>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
      />
    </svg>
  </button>
</div>
`;

exports[`Pagination basic 5`] = `
<div
  className="ts-rc-pagination"
>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
      />
    </svg>
  </button>
  <span
    className="ts-rc-pagination-pages ts-rc-pagination-pages--more"
  >
    <button
      className=""
      key="1"
      onClick={[Function]}
      type="button"
    >
      <span>
        1
      </span>
    </button>
    <button
      className=""
      key="2"
      onClick={[Function]}
      type="button"
    >
      <span>
        2
      </span>
    </button>
    <button
      className=""
      key="3"
      onClick={[Function]}
      type="button"
    >
      <span>
        3
      </span>
    </button>
    <button
      className=""
      key="4"
      onClick={[Function]}
      type="button"
    >
      <span>
        4
      </span>
    </button>
    <button
      className="ts-rc-pagination-btn--active"
      key="5"
      onClick={[Function]}
      type="button"
    >
      <span>
        5
      </span>
    </button>
  </span>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
      />
    </svg>
  </button>
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
      />
    </svg>
  </button>
</div>
`;
